### PR TITLE
3.next - Command collection autoDiscover()

### PR DESCRIPTION
**Pull request by markstory**
Build out the `autoDiscover()` method for `CommandCollection` this method will scan the core, app and all plugins for shell commands and append them into the collection. I've made the results of the scan include enough data that I can refactor the `CommandListShell` and `CompletionShell` in a subsequent pull request.

Refs #10716

### CODE REVIEW

File: src/Console/CommandCollection.php [hunk 1]
```diff
@@ -130,4 +133,62 @@ public function count()
...
+     *
+     * Commands from plugins will be added based on the order plugins are loaded.
+     * Plugin shells will attempt to use a short name. If however, a plugin
+     * provides a shell that conflicts with CakePHP or the application shells,
+     * the full `plugin_name.shell` name will be used. Plugin shells are added
+     * in the order that plugins were loaded.
+     *
+     * @return $this
+     */
+    public function autoDiscover()
```
Comment (lorenzo) [inline src/Console/CommandCollection.php]: What if autodiscover returned the list of commands instead. That would make the api look like

```php
$collection->addAll($collection->autoDiscover());
```

The benefit I see from doing this is that you could filter-out or reorder some of the results if needed.

Comment (markstory) [inline src/Console/CommandCollection.php]: Good idea. I'll make that change.

File: src/Console/CommandCollection.php [hunk 2]
```diff
@@ -60,8 +61,9 @@ public function add($name, $command)
         // Once we have a new Command class this should check
         // against that interface.
         if (!is_subclass_of($command, Shell::class)) {
+            $class = is_string($command) ? $command : get_class($command);
             throw new InvalidArgumentException(
-                "'$name' is not a subclass of Cake\Console\Shell or a valid command."
+                "Cannot use '$class' for command '$name' it is not a subclass of Cake\Console\Shell."
             );
         }
         $this->commands[$name] = $command;
```
Comment (ravage84) [inline src/Console/CommandCollection.php]: Add white line as it has nothing to do with the check above.

File: src/Console/CommandCollection.php [hunk 3]
```diff
@@ -130,4 +148,60 @@ public function count()
...
+    public function autoDiscover()
+    {
+        $scanner = new CommandScanner();
+        $shells = $scanner->scanAll();
+
+        $adder = function ($out, $shells, $key) {
+            if (empty($shells[$key])) {
+                return $out;
+            }
+            foreach ($shells[$key] as $info) {
```
Comment (ravage84) [inline src/Console/CommandCollection.php]: Add white lline for readability

File: src/Console/CommandCollection.php [hunk 4]
```diff
@@ -130,4 +148,60 @@ public function count()
...
+     * in the order that plugins were loaded.
+     *
+     * @return array An array of command names and their classes.
+     */
+    public function autoDiscover()
+    {
+        $scanner = new CommandScanner();
+        $shells = $scanner->scanAll();
+
+        $adder = function ($out, $shells, $key) {
```
Comment (ravage84) [inline src/Console/CommandCollection.php]: Why not as a protected method?
I think this method could be shorter, mor readable, may be even more self-explanatory

File: src/Console/CommandScanner.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+use Cake\Core\Configure;
+use Cake\Core\Plugin;
+use Cake\Filesystem\Folder;
+use Cake\Utility\Inflector;
+
+/**
+ * Used by CommanCollection and CommandTask to scan the filesystem
+ * for command classes.
+ *
+ * @internal
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: 👍

File: src/Console/CommandScanner.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+ */
+class CommandScanner
+{
+    /**
+     * Scan the application, cakephp and plugins for shells
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: CakePHP

Comment (ravage84) [inline src/Console/CommandScanner.php]: The order is mentiond differently, [above](https://github.com/cakephp/cakephp/pull/10767/files#diff-c4fae5f19f4310651de5ab7f45abcebfR158).

File: src/Console/CommandScanner.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            ['app']
+        );
+
+        $shellList['CORE'] = $this->scanDir(
+            dirname(__DIR__) . DIRECTORY_SEPARATOR . 'Shell' . DIRECTORY_SEPARATOR,
+            'Cake\Shell\\',
+            '',
+            ['command_list']
+        );
+        $plugins = [];
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: Add a white line above for clearer segregation

File: src/Console/CommandScanner.php [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+     * Scan the application, cakephp and plugins for shells
+     *
+     * @return array
+     */
+    public function scanAll()
+    {
+        $shellList = [];
+
+        $appNamespace = Configure::read('App.namespace');
+        $shellList['app'] = $this->scanDir(
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: extract as ``scanApp()``? Could improve reusability.

File: src/Console/CommandScanner.php [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            App::path('Shell')[0],
+            $appNamespace . '\Shell\\',
+            '',
+            ['app']
+        );
+
+        $shellList['CORE'] = $this->scanDir(
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: extract as ``scanCore()``? Could improve reusability.

File: src/Console/CommandScanner.php [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+            dirname(__DIR__) . DIRECTORY_SEPARATOR . 'Shell' . DIRECTORY_SEPARATOR,
+            'Cake\Shell\\',
+            '',
+            ['command_list']
+        );
+        $plugins = [];
+        foreach (Plugin::loaded() as $plugin) {
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: extract as ``scanPlugins()``? Could improve reusability.

File: src/Console/CommandScanner.php [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+     * @return array The list of shell info arrays based on scanning the filesystem and inflection.
+     */
+    protected function scanDir($path, $namespace, $prefix, array $hide)
+    {
+        $dir = new Folder($path);
+        $contents = $dir->read(true, true);
+        if (empty($contents[1])) {
+            return [];
+        }
+        $shells = [];
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: Add white line above for readability.

File: src/Console/CommandScanner.php [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+        foreach ($contents[1] as $file) {
+            if (substr($file, -4) !== '.php') {
+                continue;
+            }
+            $shell = substr($file, 0, -4);
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: Add white line above for readability.

File: src/Console/CommandScanner.php [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+            $name = Inflector::underscore(str_replace('Shell', '', $shell));
+            if (in_array($name, $hide, true)) {
+                continue;
+            }
+            $shells[] = [
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: Add white line above for readability.

File: src/Console/CommandScanner.php (hunk 9 above)
Comment (markstory) [inline src/Console/CommandScanner.php]: But its a protected method on an internal class. How much reuse could there be?

Comment (ravage84) [inline src/Console/CommandScanner.php]: Well, might not be that relevant, would decrease method size and increase readabilty, anyway

**Comment by lorenzo** [comment]
This looks good to me, great job!

**Comment by markstory** [comment]
Thanks @lorenzo

**Pull request merged**